src/scheduler/MpiWorld.cpp

```diff
@@ -133,9 +100,26 @@ void MpiWorld::create(const faabric::Message& call, int newId, int newSize)
         msg.set_ismpi(true);
         msg.set_mpiworldid(id);
         msg.set_mpirank(i + 1);
+        msg.set_mpiworldsize(size);
     }
 
-    sch.callFunctions(req);
+    // The scheduler returns at which host each message is scheduled
```

Comment:
The comments over the next few lines are a little hard to parse as they don't necessarily correspond to the line below and the code that builds the list of recipients is interleaved with establishing the list of ranks. 

Also remember you don't need to add markdown syntax to comments (i.e. don't need the backticks around variables), and in general it's best to avoid including variable names in comments (e.g. `rankHostMap`) as they may not get changed if someone changes the variable name.

Finally does the line setting `mutable_hosts` avoid a copy ( `*hostRankMsg.mutable_hosts() = { executedAt.begin(), executedAt.end() };`)? If not I think we usually use the syntax of `hostRankMsg.set_hosts(executedAt)` (which is a bit clearer). 

To restructure I'd suggest:

```
// Send the init messages (note that message i corresponds to rank i+1)
std::vector<std::string> executedAt = sch.callFunctions(req);

// Prepend this host for rank 0
executedAt.insert(executedAt.begin(), thisHost);

// Register host to rank mapping on this host
setHostForRank(executedAt);

// Set up list of hosts to broadcast to (excluding this host)
std::set<std::string> hosts(executedAt.begin(), executedAt.end());
hosts.erase(thisHost);

// Do the broadcast
faabric::MpiHostRankMsg ranksMsg;
ranksMsg.set_hosts(executedAt);
for (const auto& h : hosts) {
    faabric::transport::sendMpiHostRankMsg(h, hostRankMsg);
}
```

Reply:
Ok to everything besides the `set_hosts()` bit. Hosts is a `repeated` field (somewhat like a vector) and, using the `set_hosts` syntax, we'd have to add hosts one by one. This is actually how it's done for `Message` in `BatchExecRequest` ([`/faabric/src/util/func.cpp`](https://github.com/faasm/faabric/blob/2be5d88e57f1cdcdc6ec62758e7f67aab4c460d4/src/util/func.cpp#L64).

Reply:
Ah I see, sorry I was getting confused by our `inputdata` and `outputdata` fields. They are also vector-like and you can pass another vector into `set_inputdata` and `set_outputdata`, but it's because they are `bytes` fields, and not standard `repeated` fields.